File: src/main/scala/Lista3.scala
package Lista3

def last[A] (xs: List[A]): Option[A] =
    xs match {
        case Nil => None
        case x :: Nil => Some(x)
        case h :: tl => last(tl)
    }


def lastTwo[A] (xs: List[A]): Option[(A, A)] = 
    xs match {
        case Nil => None
        case x :: Nil => None
        case x :: y :: Nil => Some((x, y))
        case h :: tl => lastTwo(tl)
    }

def length[A] (xs: List[A]): Int = {
    def lengthHelper[A] (xs: List[A], n: Int): Int = 
        xs match {
            case Nil => n
            case h :: tl => lengthHelper(tl, (n + 1))
        }
    lengthHelper(xs, 0)
}

def reverse[A] (xs: List[A]): List[A] = {
    def revHelper[A] (xs: List[A], acc: List[A]): List[A] =
        xs match {
            case Nil => acc
            case h :: tl => revHelper(tl, h :: acc)
        }
    revHelper(xs, List())
}

def palindrome[A] (xs: List[A]): Boolean = reverse(xs) == xs

def contains[A] (xs: List[A], x: A): Boolean = 
    xs match {
        case Nil => false
        case h :: tl => if h == x then true else contains(tl, x)
    }

def distinct[A] (xs: List[A]): List[A] = {
    def distinctHelper[A] (xs: List[A], acc: List[A]): List[A] =
        xs match {
            case Nil => acc
            case h :: tl => if contains(acc, h) then distinctHelper(tl, acc) else distinctHelper(tl, acc :+ h)
        }
    distinctHelper(xs, List())
}

def evenIndex[A] (xs: List[A]): List[A] = {
    def evenHelper[A] (xs: List[A], acc: List[A], n: Int): List[A] =
        xs match {
            case Nil => acc
            case h :: tl => if n % 2 == 0 then evenHelper(tl, acc :+ h, n + 1) else evenHelper(tl, acc, n + 1)
        }
    evenHelper(xs, List(), 0)
}

def isPrime (x: Int): Boolean =
    x match {
        case 0 => false
        case 1 => false
        case all => {
            def primeHelper (x: Int, i: Int): Boolean =
                i match {
                    case 1 => true
                    case default => if x % i == 0 then false else primeHelper(x, default - 1)
                }
            if x < 0 then false else primeHelper(all, all - 1)
        }
    }

File: src/L/ocaml/Lista3.ml
let rec last xs = 
    match xs with
    | [] -> None
    | [x] -> Some x
    | h::tl -> last tl;;
last [1; 2; 3; 4; 5];;

let rec lastTwo xs =
  match xs with
  | [] -> None
  | [x] -> None
  | [x; y] -> Some (x, y)
  | h::tl -> lastTwo tl;;
lastTwo [1; 2; 3; 4; 5];;

let length xs =
  let rec lengthHelper xs n =
    match xs with
    | [] -> n
    | h::tl -> lengthHelper tl (n + 1) in
  lengthHelper xs 0;;
length [1; 2; 3; 4; 5];;

let reverse xs =
  let rec reverseHelper xs acc =
  match xs with
  | [] -> acc
  | h::tl -> reverseHelper tl (h::acc) in
  reverseHelper xs [];;
reverse [1; 2; 3; 4];;

let palindrome xs = reverse xs = xs;;
palindrome ["a"; "b"; "c"; "b"; "a"];;

let rec contains xs x =
  match xs with
  | [] -> false
  | h::tl -> if h == x then true else contains tl x;;
contains [1; 2;] 1;;

let distinct xs =
  let rec distinctHelper xss acc =
    match xss with
    | [] -> acc
    | h::tl -> if contains acc h then distinctHelper tl acc else distinctHelper tl (acc@[h]) in
    distinctHelper xs [];;
distinct [1; 1; 2; 3; 4; 3];;

let evenIndex xs =
  let rec evenHelper xss acc n =
    match xss with
    | [] -> acc
    | h::tl -> if n mod 2 == 0 then evenHelper tl (acc@[h]) (n + 1) else evenHelper tl acc (n + 1) in
    evenHelper xs [] 0;;
evenIndex [0; 1; 2; 3; 4; 5];;

let isPrime x =
  match x with
  | 0 -> false
  | 1 -> false
  | _ -> let rec primeHelper n i = 
    match i with
    | 1 -> true
    | _ -> if n mod i == 0 then false else primeHelper n (i - 1) in
    primeHelper x (x - 1);;
isPrime 3;;